ecommerce/models.py

```diff
@@ -0,0 +1,274 @@
+
+    @classmethod
+    def get_related_field_name(cls):
+        return "order"
+
+
+class Line(TimestampedModel):
+    """
+    A line in an Order. This contains information about a specific item which is purchased.
+    """
+
+    order = models.ForeignKey(Order, on_delete=models.PROTECT, related_name="lines")
+    product_version = models.ForeignKey(ProductVersion, on_delete=models.PROTECT)
+    quantity = models.PositiveIntegerField()
+
+    def __str__(self):
+        """Description for Line"""
+        return f"Line for order #{self.order.id}, {self.quantity} {self.product_version}(s)"
+
+
+class CouponInvoice(TimestampedModel):
+    """
+    Model for a coupon. This stores the discount for the coupon and other information about how it should be redeemed.
+
+    When a coupon is redeemed by a purchaser the counter on this object is decremented
+    and a UserCoupon object is created for that particular purchaser.
+    """
+
+    tag = models.TextField(unique=True)
+
+    def __str__(self):
+        """Description for CouponInvoice"""
+        return f"CouponInvoice {self.tag}"
+
+
+class CouponInvoiceVersion(TimestampedModel):
+    """An append-only table for CouponInvoice information"""
+
+    PROMO = "promo"
+    SINGLE_USE = "single-use"
+
+    COUPON_TYPES = [PROMO, SINGLE_USE]
+
+    invoice = models.ForeignKey(CouponInvoice, on_delete=models.PROTECT)
+
+    coupon_type = models.CharField(
+        choices=[(_type, _type) for _type in COUPON_TYPES], max_length=30
+    )
+    num_coupon_codes = models.PositiveIntegerField()
+    max_redemptions = models.PositiveIntegerField()
+    max_redemptions_per_user = models.PositiveIntegerField()
+    amount = models.DecimalField(
+        decimal_places=2,
+        max_digits=20,
+        help_text="Percent discount for a coupon. Between 0 and 1.",
+    )
+    expiration_date = models.DateTimeField(
+        null=True,
+        blank=True,
+        help_text="If set, the coupons will not be redeemable after this time",
+    )
+    activation_date = models.DateTimeField(
+        null=True,
+        blank=True,
+        help_text="If set, the coupons will not be redeemable before this time",
+    )
+    tag = models.TextField()
+
+    class Meta:
+        indexes = [models.Index(fields=["created_on"])]
+
+    def __str__(self):
+        """Description for CouponInvoiceVersion"""
+        return f"CouponInvoiceVersion for {self.num_coupon_codes} of type {self.coupon_type}"
+
+
+class Coupon(TimestampedModel):
+    """Represents a coupon code"""
+
+    coupon_code = models.CharField(max_length=50)
+    invoice = models.ForeignKey(CouponInvoice, on_delete=models.PROTECT)
+
+    def __str__(self):
+        """Description for Coupon"""
+        return f"Coupon {self.coupon_code} for {self.invoice}"
+
+
+class CouponVersion(TimestampedModel):
+    """An append-only table for coupon codes"""
+
+    coupon_code = models.CharField(max_length=50)
```

Comment:
Should this not be a foreign key to a `Coupon`

Reply:
Yes you're right